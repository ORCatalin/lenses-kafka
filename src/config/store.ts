import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import { reduxBatch } from '@manaflair/redux-batch';
import logger from 'redux-logger';

import loginReducer from '../widgets/Login/loginSlice';
import mainReducer from '../widgets/Main/mainSlice';
import chartReducer from '../widgets/Chart/chartSlice';

const reducer = {
  login: loginReducer,
  main: mainReducer,
  chart: chartReducer,
};
const middleware = [
  ...getDefaultMiddleware(),
  logger,
];
const preloadedState = {};
const store = configureStore({
  reducer,
  middleware,
  // devTools: process.env.NODE_ENV !== 'production',
  preloadedState,
  enhancers: [reduxBatch],
});

export default store;
