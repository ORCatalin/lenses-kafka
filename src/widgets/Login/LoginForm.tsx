import React from "react";
import { useDispatch, useSelector } from "react-redux";
import classnames from "classnames";
import Button from "../../components/Button";
import {
  setHost,
  setUser,
  setPassword,
  hostSelector,
  userSelector,
  passwordSelector,
  loadingStatusSelector,
  login,
  tokenSelector,
} from './loginSlice';
import { AsyncActionStatus } from "../../util/constants";

export default function Connect() {
  const dispatch = useDispatch();
  const host = useSelector(hostSelector);
  const user = useSelector(userSelector);
  const password = useSelector(passwordSelector);
  const loadingStatus = useSelector(loadingStatusSelector);
  const token = useSelector(tokenSelector);

  const btnStyle = classnames("button is-fullwidth", {
    "is-primary": true,
    "is-danger": false,
    "is-loading": loadingStatus === AsyncActionStatus.PENDING,
  });

  const onInputChange = (event) => {
    const target = event.target;
    const value = target.type === "checkbox" ? target.checked : target.value;
    const name = target.name;

    switch (name) {
      case "host":
        dispatch(setHost(value));
        break;
      case "user":
        dispatch(setUser(value));
        break;
      case "password":
        dispatch(setPassword(value));
        break;
      default:
        break;
    }
  };

  const handleLogin = () => dispatch(login());
  return (
    <nav className="panel">
      <p className="panel-heading">Connection Details</p>
      <div className="panel-block">
        <p className="control has-icons-left">
          <input
            className="input is-small"
            type="text"
            placeholder="host"
            value={host}
            name="host"
            onChange={onInputChange}
          />
          <span className="icon is-small is-left">
              <i className="fa fa-server"/>
            </span>
        </p>
      </div>
      {token && (
        <div className="panel-block">
          <p className="control has-icons-left">
            You are logged in as {user}
          </p>
        </div>
      )}
      {!token && (
        <>
          <div>
            <div className="panel-block">
              <p className="control has-icons-left">
                <input
                  className="input is-small"
                  type="text"
                  placeholder="User"
                  value={user}
                  name="user"
                  onChange={onInputChange}
                />
                <span className="icon is-small is-left">
                <i className="fa fa-user"/>
              </span>
              </p>
            </div>
            <div className="panel-block">
              <p className="control has-icons-left">
                <input
                  className="input is-small"
                  type="password"
                  placeholder="Password for Authentication"
                  value={password}
                  name="password"
                  onChange={onInputChange}
                  autoComplete="off"
                />
                <span className="icon is-small is-left">
                <i className="fa fa-lock"/>
              </span>
              </p>
            </div>
          </div>
          <div className="panel-block">
            <Button onClick={handleLogin} className={btnStyle}>
              Login
            </Button>
          </div>
        </>
      )}
    </nav>
  );
}
