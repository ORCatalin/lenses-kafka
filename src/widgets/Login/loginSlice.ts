import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import getDeepObjectValue from 'lodash.get';
import api from '../../util/api';
import { AsyncActionStatus } from '../../util/constants';

export const login = createAsyncThunk(
  'mainContainer/login',
  async (_, thunkAPI) => {
    const user = userSelector(thunkAPI.getState());
    const password = passwordSelector(thunkAPI.getState());
    const response = await api.post(`/api/login`, {user, password});
    return getDeepObjectValue(response, 'data');
  }
);

export const loginSlice = createSlice({
  name: 'login',
  initialState: {
    host: 'localhost',
    user: '',
    password: '',
    loadingStatus: AsyncActionStatus.IDLE,
    token: '',
  },
  reducers: {
    setHost: (state, action) => {
      state.host = action.payload;
    },
    setUser: (state, action) => {
      state.user = action.payload;
    },
    setPassword: (state, action) => {
      state.password = action.payload;
    },
  },
  extraReducers: {
    [login.pending]: (state) => {
      state.loadingStatus = AsyncActionStatus.PENDING;
    },
    [login.fulfilled]: (state, action) => {
      state.loadingStatus = AsyncActionStatus.FULFILLED;
      state.token = action.payload;
    },
    [login.rejected]: (state) => {
      state.loadingStatus = AsyncActionStatus.REJECTED;
    },
  }
});

export const {setHost, setUser, setPassword} = loginSlice.actions;

export const hostSelector = state => state.login.host;
export const userSelector = state => state.login.user;
export const passwordSelector = state => state.login.password;
export const loadingStatusSelector = state => state.login.loadingStatus;
export const tokenSelector = state => state.login.token;

export default loginSlice.reducer;