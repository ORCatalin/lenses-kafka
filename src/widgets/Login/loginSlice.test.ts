import store from '../../config/store';
import {setHost, hostSelector} from '../Login/loginSlice';

describe('Test login slice', () => {
  test('for setHost action', () => {
    let state = store.getState();
    expect(hostSelector(state)).toBe('localhost');
    store.dispatch(setHost('127.0.0.1'));
    expect(hostSelector(store.getState())).toBe('127.0.0.1');
    expect(store.getState().login).toMatchSnapshot();
  })
})