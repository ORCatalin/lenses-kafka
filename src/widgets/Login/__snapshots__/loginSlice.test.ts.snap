// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test login slice for setHost action 1`] = `
Object {
  "host": "127.0.0.1",
  "loadingStatus": 2,
  "password": "",
  "token": "",
  "user": "",
}
`;
